### DISCUSSION TITLE
ch1-basic/ch1-06-goroutine

### DISCUSSION THREAD

**Discussion by giscus** [chai2010/advanced-go-programming-book#606] [General]
# ch1-basic/ch1-06-goroutine

Go语言高级编程

https://chai2010.cn/advanced-go-programming-book/ch1-basic/ch1-06-goroutine.html

**Comment by HeavenTonight** [comment]
>根据 Go 语言内存模型规范，对于从无缓冲 Channel 进行的接收，发生在对该 Channel 进行的发送完成之前。因此，后台线程 <-done 接收操作完成之后，main 线程的 done <- 1 发送操作才可能完成（从而退出 main、退出程>序），而此时打印工作已经完成了。


这里的描述是不是反了？

**Reply by frankfan1990** [reply]
感觉up主的不少案例并没有经过实测，很多脑补的。（甚至是错误的结论，虽然写得确实不错，但并不严谨(甚至错误)）

**Reply by chai2010** [reply]
欢迎提交 PR 改进

**Comment by JackySu** [comment]
锁那块最好用defer mu.Unlock()不然还是有概率出问题的

**Comment by zmrgo** [comment]
最后一个例子PrimeFilter的入参类型为 :  <-chan int，实际运行发现 <-chan int 和 chan int，这两个是一回事吗

**Reply by BiteFoo** [reply]
<-chan type是read only 不能写入；chan<- type是write only，不能读；chan type是read write。

**Comment by jiangchen-123** [comment]
哎，倒数第二个例子，if i%prime != 0 的内层如果直接写out <- i会死锁的原因没有看懂……

**Comment by mlee-msl** [comment]
来一个goroutine、channel、context 的综合运用吧！！！
golang实现的并发执行多任务，并聚合收集多任务执行结果与执行状态 https://github.com/mlee-msl/taskgroup/tree/main

**Comment by mlee-msl** [comment]
**<big>来一个goroutine、channel、context 的综合运用吧！！！</big>**
golang实现的并发执行多任务，并聚合收集多任务执行结果与执行状态

`项目github地址: https://github.com/mlee-msl/taskgroup/tree/main`
